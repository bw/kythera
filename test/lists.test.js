const t = require("./util").test

describe("Lists", () => {
	t("Initialization and assignment access", `
let myList = new int[]
myList[0] = 10
	`, {compile: false})

	t("Initialization with custom type", `
let myType = obj{
	int a,
	str b,
}

let myTypeList = new myType[]	
`, {compile: false})

	t("Initialization from object list literal", `
let myTypeList = new obj{
	int a,
	str b,
}[]	
`)

	t("Access function returning list", `
let myListFn = <> int[] {
	let result = new int[]
	result[0] = 10
}

myListFn()[0]
	`, {compile: false})

	describe("List literals", () => {
		t("int", `[1, 2, 3, 4]`)
		t("bool", `[true, false]`)
		t("float", `[1.234, 3.14159]`)
		t("str", `["hello", "world"]`)
		t("null", `[null, null]`)
		t("type", `[int, float, fn<int> null]`)
		t("fn", `[<int a> null {return null}, <int b> null { return null }]`)
		t("obj", `[{a = 0, b = "asdf",}, {a = 3, b = "qwerty",}]`)
	})
})