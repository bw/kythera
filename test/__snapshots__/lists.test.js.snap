// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lists Access 1`] = `
Object {
  "index": Object {
    "kind": "literal",
    "type": "int",
    "value": 2,
  },
  "kind": "access",
  "target": Object {
    "kind": "identifier",
    "name": "myList",
  },
}
`;

exports[`Lists Access from function call result 1`] = `
Object {
  "index": Object {
    "kind": "literal",
    "type": "int",
    "value": 10,
  },
  "kind": "access",
  "target": Object {
    "arguments": Array [],
    "kind": "call",
    "target": Object {
      "kind": "identifier",
      "name": "myListFn",
    },
  },
}
`;

exports[`Lists Creation of custom type (irrelvant) 1`] = `
Object {
  "identifier": "myType",
  "kind": "let",
  "value": Object {
    "kind": "literal",
    "type": "type",
    "value": Object {
      "kind": "type",
      "origin": "builtin",
      "structure": Object {
        "a": Object {
          "kind": "type",
          "origin": "builtin",
          "type": "int",
        },
        "b": Object {
          "kind": "type",
          "origin": "builtin",
          "type": "str",
        },
      },
      "type": "obj",
    },
  },
}
`;

exports[`Lists Function returning list 1`] = `
Object {
  "identifier": "myListFn",
  "kind": "let",
  "value": Object {
    "body": Array [
      Object {
        "kind": "return",
        "value": Object {
          "kind": "new",
          "target": Object {
            "kind": "type",
            "listType": Object {
              "kind": "type",
              "origin": "builtin",
              "type": "int",
            },
            "type": "list",
          },
        },
      },
    ],
    "kind": "literal",
    "parameters": Array [],
    "returns": Object {
      "kind": "type",
      "listType": Object {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "type": "list",
    },
    "type": "fn",
  },
}
`;

exports[`Lists Initialization from object list literal 1`] = `
Object {
  "identifier": "myTypeList2",
  "kind": "let",
  "value": Object {
    "kind": "literal",
    "type": "type",
    "value": Object {
      "kind": "type",
      "listType": Object {
        "kind": "type",
        "origin": "builtin",
        "structure": Object {
          "a": Object {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
          "b": Object {
            "kind": "type",
            "origin": "builtin",
            "type": "str",
          },
        },
        "type": "obj",
      },
      "type": "list",
    },
  },
}
`;

exports[`Lists Initialization with 'let' 1`] = `
Object {
  "identifier": "myList",
  "kind": "let",
  "value": Object {
    "kind": "new",
    "target": Object {
      "kind": "type",
      "listType": Object {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "type": "list",
    },
  },
}
`;

exports[`Lists Initialization with custom type 1`] = `
Object {
  "identifier": "myTypeList",
  "kind": "let",
  "value": Object {
    "kind": "new",
    "target": Object {
      "kind": "type",
      "listType": Object {
        "kind": "type",
        "name": "myType",
        "origin": "named",
      },
      "type": "list",
    },
  },
}
`;
